### PR TITLE
 wake-format documentation

### PR DESCRIPTION
**Pull request by V-FEXrt**
This is the first draft of the high level documentation for `wake-format`. It should include most things needed to jump in and contribute to `wake-format`

### CODE REVIEW

File: tools/wake-format/README.md
Comment (mmjconolly) [inline tools/wake-format/README.md:21]: > or `./bin/wake build default` from within

This implies that they've already run `make` at least once, it might be nice to add a Makefile target for the formatter even if it just calls wake under the hood

File: tools/wake-format/README.md
Comment (mwachs5) [inline tools/wake-format/README.md:5]:
```suggestion
## Usage
```

should there be some heading levels?

File: tools/wake-format/README.md
Comment (mmjconolly) [inline tools/wake-format/README.md:23]: We should probably have the build system produce a normalized name for it, either via the makefile, or via wake's `installAs`

File: tools/wake-format/README.md
Comment (mmjconolly) [inline tools/wake-format/README.md:31]: > current CST node

It may be helpful to spell out the acronym as individual words the first time it's used

File: tools/wake-format/README.md
Comment (mmjconolly) [inline tools/wake-format/README.md:86]: is `"foo "` intended to have the trailing space char? when I look at `next()` the `id` seems to be separated from spaces around it

Comment (V-FEXrt) [reply tools/wake-format/README.md:86]: The trailing space comes from `space()` on line 84. I think I was trying to make the example slightly less trivial however if it is confusing it should just be removed

File: tools/wake-format/README.md
Comment (V-FEXrt) [reply tools/wake-format/README.md:5]: Yeah, I need to figure out the correct nesting. I hit `#####` at one point which I couldn't tell apart from normal text so I un-nested everything by one, I don't think that was the correct choice though. Will work on it! Thanks :)

File: tools/wake-format/README.md
Comment (mmjconolly) [inline tools/wake-format/README.md:63]: Is it important that the user knows that it's a `uint8_t`, wondering if a typedef (or other opaque) would make it more self describing

File: tools/wake-format/README.md
Comment (mmjconolly) [reply tools/wake-format/README.md:86]: oh somehow my eyes entirely slid over the `space()`

File: tools/wake-format/README.md
Comment (V-FEXrt) [reply tools/wake-format/README.md:63]: Jake and I talked about this previously. I'm actually very pro adding a typedef here. Would make it easier to bump the size if we ever have more that 256 IDs as well. 

It's one of those things where I'd want to change it everywhere if we added it though. I'll look into at least making the change for `wake-format` but I think `wake` would greatly benefit from it

File: tools/wake-format/README.md
Comment (mmjconolly) [inline tools/wake-format/README.md:119]: When we do `doc.as_string()` and get `foo`, is that in replacement for the entierity of `id = 5`, or just `id`, or all of `def id = 5`?

File: tools/wake-format/README.md
Comment (JakeSiFive) [reply tools/wake-format/README.md:23]: Yeah we just add it to the installs with shim-wake, job-cache, and wakebox

File: tools/wake-format/README.md
Comment (JakeSiFive) [reply tools/wake-format/README.md:31]: It should be noted that this documentation is largely intended for wake devs at this time, we'll add user documentation once we have a user interface :)

File: tools/wake-format/README.md
Comment (JakeSiFive) [reply tools/wake-format/README.md:119]: Well this example would actually crash because it didn't consume all the nodes.  Ignoring that this would only output "foo" because only the `def` was consumed.

File: tools/wake-format/README.md
Comment (mmjconolly) [reply tools/wake-format/README.md:31]: Yep I can see that, mostly I myself had forgotten what CST stood for

File: tools/wake-format/README.md
Comment (JakeSiFive) [reply tools/wake-format/README.md:119]: There's a partial analogy for this being dual to a parser/regex if that's helpful for you. `format` implicitly has an ending anchor at the end which is why this would fail. But ignoring that anchor, this would only match and consume the `def` and would output `foo` in its place.

File: tools/wake-format/README.md
Comment (mmjconolly) [inline tools/wake-format/README.md:413]: Is this regarding things like line length?

File: tools/wake-format/README.md
Comment (V-FEXrt) [reply tools/wake-format/README.md:119]: Yeah I was actually debating how to handle the fact that most of these examples actually crash. Fully consuming the input makes the example tedious, and giving a partial node is unclear.

maybe an imaginary `consume_rest()` function

Comment (JakeSiFive) [reply tools/wake-format/README.md:119]: walk is also annoying to deal with in examples like this but unavoidable I think

File: tools/wake-format/README.md
Comment (V-FEXrt) [reply tools/wake-format/README.md:21]: https://github.com/sifive/wake/pull/975 should fix that, I'll update the documentation once it lands

File: tools/wake-format/README.md
Comment (V-FEXrt) [reply tools/wake-format/README.md:23]: If you run `make install` then it _does_ get installed/renamed appropriately

```
> DESTDIR=../install make install
WAKE_PATH=/usr/bin ./bin/wake build default
Pass "BUILD"
WAKE_PATH=/usr/bin ./bin/wake install ../install
Pass "INSTALL"
> ls ../install/bin 
job-cache  wake  wakebox  wake-format
```

Is that what you mean, or do we also want to copy from `./bin/wake-format.native-cpp14-release` to `./bin/wake-format`

File: tools/wake-format/README.md
Comment (JakeSiFive) [reply tools/wake-format/README.md:21]: I think users that would prefer `make wake-format` can just call `make` instead.

File: tools/wake-format/README.md
Comment (JakeSiFive) [reply tools/wake-format/README.md:23]: That's all that's needed. I think you've satisfied the requirement.

File: tools/wake-format/README.md
Comment (JakeSiFive) [inline tools/wake-format/README.md:166]: pred_fits

File: tools/wake-format/README.md
Comment (JakeSiFive) [inline tools/wake-format/README.md:196]: Maybe mention the WALK macro here?

**Review by JakeSiFive — APPROVED** [review]
I'm quite happy with this but maybe ask for an overview from Matthew/Sam or Wes to get a second pair of eyes. I'll try to write at least one of those TODO sections a week from here.

**Comment by JakeSiFive** [comment]
Ashley said I could close this

**Pull request closed without merging**
